pkg/controller/nodeobservatory/controller.go

```diff
@@ -0,0 +1,397 @@
+			n.queue.AddAfter(key, BaseDelay)
+		}
+
+		return true
+	}
+
+	if n.queue.NumRequeues(key) < 5 {
+		// Re-enqueue the key rate limited. Based on the rate limiter on the
+		// queue and the re-enqueue history, the key will be processed later again.
+		n.queue.AddRateLimited(key)
+		return true
+	}
+
+	// Retries exceeded. Forgetting for this reconciliation loop
+	n.queue.Forget(key)
+	return true
+}
+
+func (n *NodeObservatory) reconcile(kluster *v1.Kluster) (bool, error) {
+
+	n.cleanUpInformers()
+
+	if kluster != nil {
+		if err := n.createAndWatchNodeInformerForKluster(kluster); err != nil {
+			return true, err
+		}
+	}
+
+	return false, nil
+}
+
+func (n *NodeObservatory) cleanUpInformers() {
+	n.nodeInformerMap.Range(
+		func(key, value interface{}) bool {
+			if _, exists, _ := n.Factories.Kubernikus.Kubernikus().V1().Klusters().Informer().GetIndexer().GetByKey(key.(string)); !exists {
+				if i, ok := n.nodeInformerMap.Load(key); ok {
+					informer := i.(*NodeInformer)
+					informer.close()
+				}
+				n.nodeInformerMap.Delete(key)
+			}
+			return true
+		},
+	)
+}
+
+func (n *NodeObservatory) createAndWatchNodeInformerForKluster(kluster *v1.Kluster) error {
+	key, err := cache.MetaNamespaceKeyFunc(kluster)
+	if err != nil {
+		return err
+	}
+
+	if _, exists := n.nodeInformerMap.Load(key); !exists {
+		n.logger.Log(
+			"creating nodeInformer for kluster %s", key,
+			"v", 2,
+		)
+		nodeInformer, err := newNodeInformerForKluster(n.Clients.Satellites, kluster, make(chan struct{}))
+		if err != nil {
+			return err
+		}
+
+		nodeInformer.SharedIndexInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				if kluster, err = n.getKlusterByKey(key); err != nil {
+					n.logger.Log(err)
+					return
+				}
+				n.handlersMux.RLock()
+				defer n.handlersMux.RUnlock()
+				for _, addHandler := range n.addEventHandlers {
+					addHandler(kluster, obj.(*api_v1.Node))
+				}
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				if kluster, err = n.getKlusterByKey(key); err != nil {
+					n.logger.Log(
+						"err", err,
+						"v", 2,
+					)
+					return
+				}
+				n.handlersMux.RLock()
+				defer n.handlersMux.RUnlock()
+				for _, updateHandler := range n.updateEventHandlers {
+					updateHandler(kluster, oldObj.(*api_v1.Node), newObj.(*api_v1.Node))
+				}
+			},
+			DeleteFunc: func(obj interface{}) {
+				if kluster, err = n.getKlusterByKey(key); err != nil {
+					n.logger.Log(
+						"err", err,
+						"v", 2,
+					)
+					return
+				}
+				n.handlersMux.RLock()
+				defer n.handlersMux.RUnlock()
+				for _, deleteHandler := range n.deleteEventHandlers {
+					deleteHandler(kluster, obj.(*api_v1.Node))
+				}
+			},
+		})
+
+		n.logger.Log(
+			"storing nodeInformer for kluster %v", key,
+			"v", 1,
+		)
+		n.nodeInformerMap.Store(
+			key,
+			nodeInformer,
+		)
+
+		go func(informer *NodeInformer) {
+			n.informerWg.Add(1)
+			defer n.informerWg.Done()
+			for {
```

Comment:
I'm not sure this for loop works as intended. They way I see it the go routine get stuck at `informer.run()` and never gets to wait for `n.stopCh` to close.